sdk/src/Core/Amazon.Runtime/Internal/Util/StreamingUtf8JsonReader.cs

```diff
@@ -0,0 +1,183 @@
+            // supported in C# 11+
+            _reader = new Utf8JsonReader(_buffer);
+            HandleUtf8Bom(ref _buffer);
+        }
+
+        private void HandleUtf8Bom(ref byte[] buffer)
+        {
+            int utf8BomLength = JsonConstants.Utf8Bom.Length;
+            Debug.Assert(buffer.Length >= utf8BomLength);
+
+            int bytesRead = _stream.Read(buffer, 0, buffer.Length);
+            int start = 0;
+            if (buffer.AsSpan().StartsWith(JsonConstants.Utf8Bom))
+            {
+                start += utf8BomLength;
+                bytesRead -= utf8BomLength;
+            }
+            _reader = new Utf8JsonReader(buffer.AsSpan(start, bytesRead), isFinalBlock: bytesRead == 0, default);
+        }
+
+        // Custom delegate to handle ref parameters
+        public delegate void RefAction(ref Utf8JsonReader reader);
+
+        /// <summary>
+        /// Method to allow passing the private _reader as ref. This is a hacky way to get around the error
+        /// "a property or indexer may not be passed as an out or ref parameter". Use this when you need to pass
+        /// the Utf8JsonReader to a method that requires a ref parameter. <see cref="Amazon.Runtime.Internal.Transform.JsonUnmarshallerContext.ToJsonDocument(ref StreamingUtf8JsonReader)"/>
+        /// for usage example.
+        /// </summary>
+        /// <param name="action">The custom delegate</param>
+        public void PassReaderByRef(RefAction action)
+        {
+            action(ref _reader);
+        }
+
+        /// <summary>
+        /// Reads data from the buffer and fetches more data from the stream to fill the buffer it more data exists.
+        /// </summary>
+        /// <returns>true if there is more data, false otherwise.</returns>
+        public bool Read()
+        {
+            // hasMoreData can return false if the value starts in one buffer and leaks into the next buffer
+            bool hasMoreData = _reader.Read();
+
+            while (!hasMoreData)
+            {
+                GetMoreBytesFromStream(_stream, ref _buffer, ref _reader);
+
+                if (_reader.IsFinalBlock)
+                    break;
+
+                hasMoreData = _reader.Read();
+            }
+
+            return hasMoreData;
+        }
+
+        private static void GetMoreBytesFromStream(Stream stream, ref byte[] buffer, ref Utf8JsonReader reader)
+        {
+            int bytesRead = 0;
+            bool resized = false;
+            int previousBufferLength = buffer.Length;
+
+            if (reader.BytesConsumed < buffer.Length)
+            {
+                // if Read() returned false and we are here inside this if block that means that we couldn't fully parse the JSON token 
+                // because it was too large to fit in the remainder of the buffer.
+                ReadOnlySpan<byte> leftover = buffer.AsSpan().Slice((int)reader.BytesConsumed);
+
+                if (leftover.Length == buffer.Length)
```

Comment:
So this basically covers the case where on one read  BytesConsumed is less than buffer.Length, meaning that there is a JSON token that can't be fully read b/c it ends in the next part of the stream. So we copy the leftover to the current buffer and fill the buffer with the remainder of the stream. Then on the NEXT read if leftover.Length == buffer.Length that means 0 data was read, which means even after copying over the leftover from the previous read and reading more data in, we still can't read the full JSON token so we must double the buffer length to see if that would give the reader enough space to read the JSON token.

Reply:
Okay but the comment about the `if` block is over the `leftover` and logic about going into the if block is more about if `BytesConsumed` is equal to 0. It is an extra layer of logic to understand that if `BytesConsumed` is equal to 0 then `leftover.Length == buffer.Length` is true. The logic flow is also hard to read in general because the conditions are spread out across the function with more variables being created that are only used sometimes. For example if you are recreating the buffer you go in the first `if` block then skip some code then go in another blocking continue the resize logic if the `resized` variable is true.

This is my attempt trying to keep the related logic close to each other and get rid of excess variables. I haven't test this though.

```csharp
int bytesRead = 0;

if (reader.BytesConsumed < buffer.Length)
{
    // If BytesConsumed is 0 then the previous reader.Read() method failed because
    // the next JSON token to read from the stream is too large to fit in the buffer. 
    // That means we need to resize the buffer and try again to read the next JSON token.
    if (reader.BytesConsumed == 0)
    {                        
        // Rent a new buffer twice the size of the previous and fill the new resized buffer with data from the stream
        var resizedBuffer = ArrayPool<byte>.Shared.Rent(Math.Min(int.MaxValue, (buffer.Length * 2)));
        bytesRead = FillBuffer(stream, ref resizedBuffer, 0, resizedBuffer.Length);

        // Recreate ref parameter reader using a span of the data from the new resized buffer.
        var resizedSpan = resizedBuffer.AsSpan(0, bytesRead + buffer.Length);
        reader = new Utf8JsonReader(resizedSpan, isFinalBlock: bytesRead == 0, reader.CurrentState);

        // Return old buffer and update the ref buffer parameter to new resized buffer.
        Logger.GetLogger(typeof(StreamingUtf8JsonReader)).DebugFormat("Resizing buffer from {0} to {1}", buffer.Length, resizedBuffer.Length);
        ArrayPool<byte>.Shared.Return(buffer);
        buffer = resizedBuffer;

        // Do an early return since we have a new Utf8JsonReader with a filled buffer.
        return;
    }
    else
    {
        // Move the unprocessed data from the buffer to the start and the fill
        // remaining space in the buffer with new content from the stream.
        ReadOnlySpan<byte> leftover = buffer.AsSpan().Slice((int)reader.BytesConsumed);
        leftover.CopyTo(buffer);
        bytesRead = FillBuffer(stream, ref buffer, leftover.Length, buffer.Length - leftover.Length);
    }
}

else
{
    bytesRead = FillBuffer(stream, ref buffer, 0, buffer.Length);
}
```